File: frontend_rebuild/types/listing.ts
import type { UserRole } from "@/types/auth";
import type { AreaRead, CampusRead } from "@/types/location";

export type ListingType = "rental" | "buy_property" | "marketplace";

export type ListingStatus =
  | "draft"
  | "pending"
  | "published"
  | "rejected"
  | "expired"
  | "sold"
  | "rented"
  | "archived";

export type MediaType = "image" | "video";

export type RentalCategory =
  | "hostel"
  | "single_room"
  | "self_contain"
  | "shared_apartment"
  | "mini_flat"
  | "flat"
  | "duplex";

export type RentPeriod = "monthly" | "six_months" | "yearly";

export type PropertyCategory =
  | "land"
  | "house"
  | "duplex"
  | "commercial"
  | "office"
  | "warehouse";

export type PropertyCondition = "new" | "good" | "renovation_required";

export type MarketplaceCategory =
  | "phones"
  | "laptops"
  | "furniture"
  | "electronics"
  | "appliances"
  | "fashion"
  | "books"
  | "gadgets"
  | "services"
  | "others";

export type ItemCondition = "new" | "like_new" | "used" | "fair";

export interface PublicUser {
  id: string;
  full_name: string;
  role: UserRole;
  profile_image_url: string | null;
}

export interface ListingMedia {
  id: string;
  media_type: MediaType;
  url: string;
  sort_order: number;
  is_cover: boolean;
}

export interface RentalDetails {
  category: RentalCategory;
  rent_period: RentPeriod;
  bedrooms: number | null;
  bathrooms: number | null;
  toilets: number | null;
  is_furnished: boolean;
  caution_fee: number | string | null;
  service_charge: number | string | null;
  distance_to_campus_km: number | string | null;
  address: string;
}

export interface BuyPropertyDetails {
  category: PropertyCategory;
  property_condition: PropertyCondition;
  bedrooms: number | null;
  bathrooms: number | null;
  toilets: number | null;
  land_size_sqm: number | string | null;
  title_document: string | null;
  address: string;
}

export interface MarketplaceDetails {
  category: MarketplaceCategory;
  condition: ItemCondition;
  is_negotiable: boolean;
}

export interface Listing {
  id: string;
  owner: PublicUser;
  area: AreaRead;
  campus: CampusRead | null;
  listing_type: ListingType;
  title: string;
  slug: string;
  description: string;
  price: number | string;
  currency: string;
  status: ListingStatus;
  rejection_reason: string | null;
  is_featured: boolean;
  featured_until: string | null;
  published_at: string | null;
  expires_at: string | null;
  view_count: number;
  contact_count: number;
  media: ListingMedia[];
  rental_details: RentalDetails | null;
  buy_property_details: BuyPropertyDetails | null;
  marketplace_details: MarketplaceDetails | null;
  created_at: string;
  updated_at: string;
}

export interface ListingFilters {
  listing_type?: ListingType;
  area_id?: string;
  min_price?: number;
  max_price?: number;
  q?: string;
  limit?: number;
  offset?: number;
}

export interface SavedListing {
  listing: Listing;
  created_at: string;
}

export type InquiryType = "general" | "inspection";

export interface InquiryCreatePayload {
  inquiry_type: InquiryType;
  message: string;
}

export interface ListingMediaInput {
  media_type: MediaType;
  url: string;
  storage_public_id?: string | null;
  sort_order?: number;
  is_cover?: boolean;
}

export interface ListingBaseCreatePayload {
  area_id: string;
  campus_id?: string | null;
  title: string;
  description: string;
  price: number;
  media: ListingMediaInput[];
}

export interface MarketplaceCreatePayload extends ListingBaseCreatePayload {
  category: MarketplaceCategory;
  condition: ItemCondition;
  is_negotiable: boolean;
}

export interface RentalCreatePayload extends ListingBaseCreatePayload {
  category: RentalCategory;
  rent_period: RentPeriod;
  bedrooms?: number | null;
  bathrooms?: number | null;
  toilets?: number | null;
  is_furnished: boolean;
  caution_fee?: number | null;
  service_charge?: number | null;
  distance_to_campus_km?: number | null;
  address: string;
}

export interface BuyPropertyCreatePayload extends ListingBaseCreatePayload {
  category: PropertyCategory;
  property_condition: PropertyCondition;
  bedrooms?: number | null;
  bathrooms?: number | null;
  toilets?: number | null;
  land_size_sqm?: number | null;
  title_document?: string | null;
  address: string;
}

File: backend_rebuild/app/schemas/listing.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, HttpUrl

from app.models.enums import (
    ItemCondition,
    ListingStatus,
    ListingType,
    MarketplaceCategory,
    MediaType,
    PropertyCategory,
    PropertyCondition,
    RentalCategory,
    RentPeriod,
)
from app.schemas.common import ORMModel
from app.schemas.location import AreaRead, CampusRead
from app.schemas.user import PublicUserRead


class ListingMediaInput(BaseModel):
    media_type: MediaType
    url: HttpUrl
    storage_public_id: str | None = Field(default=None, max_length=255)
    sort_order: int = Field(default=0, ge=0)
    is_cover: bool = False


class ListingMediaRead(ORMModel):
    id: UUID
    media_type: MediaType
    url: str
    sort_order: int
    is_cover: bool


class ListingBaseCreate(BaseModel):
    area_id: UUID
    campus_id: UUID | None = None
    title: str = Field(min_length=5, max_length=220)
    description: str = Field(min_length=30, max_length=12000)
    price: Decimal = Field(ge=0, max_digits=14, decimal_places=2)
    media: list[ListingMediaInput] = Field(default_factory=list)


class RentalCreate(ListingBaseCreate):
    category: RentalCategory
    rent_period: RentPeriod
    bedrooms: int | None = Field(default=None, ge=0, le=100)
    bathrooms: int | None = Field(default=None, ge=0, le=100)
    toilets: int | None = Field(default=None, ge=0, le=100)
    is_furnished: bool = False
    caution_fee: Decimal | None = Field(default=None, ge=0)
    service_charge: Decimal | None = Field(default=None, ge=0)
    distance_to_campus_km: Decimal | None = Field(default=None, ge=0)
    address: str = Field(min_length=5, max_length=2000)


class BuyPropertyCreate(ListingBaseCreate):
    category: PropertyCategory
    property_condition: PropertyCondition
    bedrooms: int | None = Field(default=None, ge=0, le=100)
    bathrooms: int | None = Field(default=None, ge=0, le=100)
    toilets: int | None = Field(default=None, ge=0, le=100)
    land_size_sqm: Decimal | None = Field(default=None, ge=0)
    title_document: str | None = Field(default=None, max_length=180)
    address: str = Field(min_length=5, max_length=2000)


class MarketplaceCreate(ListingBaseCreate):
    category: MarketplaceCategory
    condition: ItemCondition
    is_negotiable: bool = False


class ListingRejectRequest(BaseModel):
    reason: str = Field(min_length=10, max_length=2000)


class RentalDetailsRead(ORMModel):
    category: RentalCategory
    rent_period: RentPeriod
    bedrooms: int | None
    bathrooms: int | None
    toilets: int | None
    is_furnished: bool
    caution_fee: Decimal | None
    service_charge: Decimal | None
    distance_to_campus_km: Decimal | None
    address: str


class BuyPropertyDetailsRead(ORMModel):
    category: PropertyCategory
    property_condition: PropertyCondition
    bedrooms: int | None
    bathrooms: int | None
    toilets: int | None
    land_size_sqm: Decimal | None
    title_document: str | None
    address: str


class MarketplaceDetailsRead(ORMModel):
    category: MarketplaceCategory
    condition: ItemCondition
    is_negotiable: bool


class ListingRead(ORMModel):
    id: UUID
    owner: PublicUserRead
    area: AreaRead
    campus: CampusRead | None
    listing_type: ListingType
    title: str
    slug: str
    description: str
    price: Decimal
    currency: str
    status: ListingStatus
    rejection_reason: str | None
    is_featured: bool
    featured_until: datetime | None
    published_at: datetime | None
    expires_at: datetime | None
    view_count: int
    contact_count: int
    media: list[ListingMediaRead]
    rental_details: RentalDetailsRead | None
    buy_property_details: BuyPropertyDetailsRead | None
    marketplace_details: MarketplaceDetailsRead | None
    created_at: datetime
    updated_at: datetime